### ISSUE TITLE
Discussion: should "clone" of a constant object return a new mutable object?

### ISSUE THREAD

**Issue by mirguest** [AIDASoft/podio#219]
In the current version, the clone method of a constant object returns a constant object. But is that possible to return a mutable object?
In my simulation, a part of MC particles are from "Generator", while another part are from "Detector simulation". Due to the const object, I need to create two collections of MC particles for "Generator" and "Detector Simulation". Then I want to copy the content of MC particle from "Generator" to "Detector simulation". If there is a method could help me fill the values automatically, it would be more convenient. 
Or is that any other way to solve my problem?
Thank you!
Tao

**Comment by tmadlener** [comment]
Hi Tao, I ran into the same thing as you yesterday, and I think you are right: a clone of a Const object should be mutable, since the cloned object is completely separate from the original one. Would making `clone` always return a mutable object solve your problem already, or would you prefer to also have a function that clones a complete collection?

**Comment by gaede** [comment]
Hi. I in principle agree that a `clone` method needs to return a mutable object - after all it is a deep copy of the original and hence a new object. However, this would not solve Tao's problem, as the parent and daughter relations would still point to the original collection and this is clearly not what is needed here. This shows the general problem of a `clone` method with relations: should these point to the original objects, which could be useful eg. for Tracks or Clusters or should they point to new - cloned - versions of the pointed to objects, which in turn would also not work for the case of MCParticles ?
Probably we need to revisit the issue of having a `clone` method at all - and if how it should be implemented. (In LCIO we never found a solution to this). 
Concerning the copy of the MCParticle collection: I guess you would have to write your own utility that copies the original collection element wise and restores the parent daughter relations (e.g. via some smart maps of parent and daughter indices from the original collection)

**Comment by mirguest** [comment]
Hi Thomas and Frank,
Thank you for your comments. 
That's true that there will be a problem if there are relation in the MC particles. Currently, I create a new collection for the MC particles during detector simulation. In this new collection, I create new objects and set the values from the const objects, no relation are stored. Then the secondaries are also created in the new collection and then I setup the relations. 
My change in https://github.com/cepc/CEPCSW/pull/185: https://github.com/cepc/CEPCSW/pull/185/files#diff-8b9f430562d469e820f3f22d47277b0ff2e2f7411bc2d4281b6061114f7d1169
Tao

**Comment by tmadlener** [comment]
That is a good point. I didn't really think of the relations and the different possibilities there.

> This shows the general problem of a clone method with relations: should these point to the original objects, which could be useful eg. for Tracks or Clusters or should they point to new - cloned - versions of the pointed to objects, which in turn would also not work for the case of MCParticles ?

I think having a clone version that points to new versions of the related to objects is not something that can be solved in podio, but would rather have to be addressed via some utility in the EDM itself. The reason for this is, that there is an in principle arbitrary number of relations for each object, which also means that there is an arbitrary number of collections involved. Additionally, there might be cases where only some of the relations should be "cloned", while others should remain.

A possibility could be to offer a `clone` method that takes a flag argument. Depending on the value of that it will only copy the data, and initialize the relations to point to nothing, or copy the data and keep the relations the same as they were in the cloned-from object (i.e. they still point to the same objects as the relations of the cloned-from objects). This would at least make it easy to copy the data over (without having to risk some "divergences" during schema evolution), and would still allow to keep the relations if wanted. If some more sophisticated relation handling is necessary, some dedicated code is probably necessary in any case.

**Comment by hegner** [comment]
Let me comment here as well, copying many of the things you already said:

- yes, a clone should be mutable (need to dig in the history why it isn't)
- a deep copy of objects will not work as it will involve objects of different collections, which would need to be cloned etc.
- having different flags for clone would not be helpful, as with a mutable object the user has all the freedom to do what is needed, including resetting etc. We would ever only be able to cover a tiny part of needs with flags. So in my opinion we would only increase the confusion about the API
- there is however nothing that prevents us from providing some higher level functionality via external functors to do more complicated operations. The input to that however would likely not be a single object...
- We should keep the core API simple

**Comment by tmadlener** [comment]
Just for completeness: #220 would make the `clone` implementation return a mutable object that has all the data copied and keeps the same relations as the original object.

**Comment by vvolkl** [comment]
So I also stumbled across this issue while trying to update downstream code for the new interfaces. There is absolutely a need to get a mutable object from a const one But  "clone" is, at least in my understanding, an *identical* copy of the object -- I would find it very confusing if what I cloned is suddenly mutable.

So rather than change clone, how about adding `copy()` or `mutableCopy()` with the desired behavior? Added benefit that existing interfaces are not changed, which should always be done with loads of care!

**Comment by tmadlener** [comment]
> But "clone" is, at least in my understanding, an identical copy of the object -- I would find it very confusing if what I cloned is suddenly mutable.

I am probably missing something here, but if you want an *identical* immutable copy of the object, but what would be the benefits of having a "real" clone instead of just using an (implicit) copy of the original object? Cloning could be used for some sort of "lifetime extension" for the underlying data, but the relations might have different lifetime(s), so the object you get is not really *identical* in that case.

**Comment by vvolkl** [comment]
No, my point is that the `clone` name is wrong for creating a mutable copy. To me it is more intuitive to keep clone as it is ( although I also can't think of many uses)  and add something like `mutableCopy()` or a ctor of a mutable object that takes a const one.

**Comment by tmadlener** [comment]
I see your point. There is just one thing I think we should avoid: Having different names for this functionality on the mutable and immutable objects. We at least try to hide the fact that we have two different user facing types in other places, and I think it would be a pity if we would "highlight" a difference between the two for this functionality.

**Comment by gaede** [comment]
I think it would actually be cleaner and clearer to have a clone method if a new object is created. The copy c‘tors are used in our value semantics to refer to an existing object be it mutable or not…

**Comment by gaede** [comment]
@tmadlener had not seen your previous comment. But yes I think this actually is an argument ;–)

**Comment by tmadlener** [comment]
> @tmadlener had not seen your previous comment. But yes I think this actually is an argument ;–)

I think our argument is the same. I just meant that "having to document something" is not an argument against a solution in this case, because that is something that we have to do for all solutions.

**Comment by hegner** [comment]
I would stay with a clone method, that returns a mutable. No copy constructors please. As there may be a lot of implicit behaviour in user code, argument passing etc if used wrongly. A 'clone' is just much more simple

And yes - documentation is the key.

**Comment by vvolkl** [comment]
Ok, there seems to be a consensus here -- fine by me, although I still think that the consequence here is that users will encounter code like:
```
auto a = b.clone();
```
that I find very ambiguous, while
```
auto a = b.mutableCopy();
```
or something idiomatic like a copy constructor (that definitely would have to be explicit) makes it clearer what's going on without needing to refer to the docs.
